daemon/response.go

```diff
@@ -281,55 +282,137 @@ func (f fileResponse) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+}
+
+func (rr *journalLineReaderSeqResponse) logReader(r io.ReadCloser, c chan systemd.Log, e chan error) {
+	defer r.Close()
+	decoder := json.NewDecoder(r)
+	for {
+		var log systemd.Log
+		if err := decoder.Decode(&log); err != nil {
+			e <- err
+			break
+		}
+		c <- log
+	}
 }
 
 func (rr *journalLineReaderSeqResponse) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 	w.Header().Set("Content-Type", "application/json-seq")
 
 	flusher, hasFlusher := w.(http.Flusher)
-
-	var err error
-	dec := json.NewDecoder(rr)
 	writer := bufio.NewWriter(w)
 	enc := json.NewEncoder(writer)
-	for {
-		var log systemd.Log
-		if err = dec.Decode(&log); err != nil {
-			break
-		}
 
-		writer.WriteByte(0x1E) // RS -- see ascii(7), and RFC7464
+	// start all the readers
+	c := make(chan systemd.Log)
+	e := make(chan error)
+	for _, r := range rr.readers {
+		go rr.logReader(r, c, e)
+	}
 
-		// ignore the error...
-		t, _ := log.Time()
-		if err = enc.Encode(client.Log{
-			Timestamp: t,
-			Message:   log.Message(),
-			SID:       log.SID(),
-			PID:       log.PID(),
-		}); err != nil {
-			break
+	writeError := func(err error) {
+		fmt.Fprintf(writer, `\x1E{"error": %q}\n`, err)
+		logger.Noticef("cannot stream response; problem reading: %v", err)
+	}
+
+	writeLogs := func(logs []systemd.Log) error {
+		// sort by timestamp ascending
+		sort.Slice(logs, func(i, j int) bool {
+			ti, _ := logs[i].Time()
+			tj, _ := logs[j].Time()
+			return ti.Before(tj)
+		})
+
+		for _, l := range logs {
+			writer.WriteByte(0x1E) // RS -- see ascii(7), and RFC7464
+
+			// ignore the error...
+			t, _ := l.Time()
+			if err := enc.Encode(client.Log{
+				Timestamp: t,
+				Message:   l.Message(),
+				SID:       l.SID(),
+				PID:       l.PID(),
+			}); err != nil {
+				return err
+			}
 		}
 
 		if rr.follow {
-			if e := writer.Flush(); e != nil {
-				break
+			if err := writer.Flush(); err != nil {
+				return nil
 			}
 			if hasFlusher {
 				flusher.Flush()
 			}
 		}
+		return nil
 	}
-	if err != nil && err != io.EOF {
-		fmt.Fprintf(writer, `\x1E{"error": %q}\n`, err)
-		logger.Noticef("cannot stream response; problem reading: %v", err)
+
+	var logReadersDone int
+	for {
+		var logs []systemd.Log
+
+		// always block read on the first one to ensure we don't waste any
+		// time spinning on a channel that will never have any logs
+		select {
+		case log := <-c:
+			logs = append(logs, log)
+		case err := <-e:
+			if err != io.EOF {
+				writeError(err)
+			}
+			logReadersDone++
+		}
+
+		// Now we spend a small amount of time batch reading all available logs
+		// on the log/error channel, the reason we do this is to make sure we read
+		// everything available for the initial batch of logs (when following) or
+		// we read everything available when not following. It's done like this to
+		// ensure that output is consistent and sorted correctly by timestamp when
+		// multiple log-sources are in play.
+		timeout := time.After(time.Millisecond * 25)
```

Comment:
This might be a tad controversal, and also the number of 25 milliseconds is arbitrarily chosen to just provide a small timeout for batch reading. This is done for two reasons, the first reason being we want to correctly sort multiple log sources by reading the entirety of the initial provided log entries when following, but also when the entire log is available when not following. The second reason being here that we don't want to get stuck reading logs if they get spammed over longer intervals (being >25ms in this case, we can adjust this).